state/enableha_test.go

```diff
@@ -347,6 +348,67 @@ func (s *EnableHASuite) TestEnableHADefaultsTo3(c *gc.C) {
 	c.Assert(m3.IsManager(), jc.IsTrue)
 }
 
+// progressControllerToDead starts the machine as dying, and then does what the normal workers would do
+// (like peergrouper), and runs all the cleanups to progress the machine all the way to dead.
+func (s *EnableHASuite) progressControllerToDead(c *gc.C, id string) {
+	m, err := s.State.Machine(id)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Logf("destroying machine 0")
+	c.Assert(m.Destroy(), jc.ErrorIsNil)
+	c.Assert(m.Refresh(), jc.ErrorIsNil)
+	c.Check(m.WantsVote(), jc.IsFalse)
+	// Pretend to be the peergrouper, notice the machine doesn't want to vote, so get rid of its vote, and remove it
+	// as a controller machine.
+	m.SetHasVote(false)
+	c.Assert(s.State.RemoveControllerMachine(m), jc.ErrorIsNil)
+	c.Assert(s.State.Cleanup(), jc.ErrorIsNil)
+	c.Assert(m.EnsureDead(), jc.ErrorIsNil)
+}
+
+func (s *EnableHASuite) TestEnableHAGoesToNextOdd(c *gc.C) {
+	changes, err := s.State.EnableHA(0, constraints.Value{}, "quantal", nil, "0")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(changes.Added, gc.HasLen, 3)
+	s.assertControllerInfo(c, []string{"0", "1", "2"}, []string{"0", "1", "2"}, nil)
+	s.PatchValue(state.ControllerAvailable, func(m *state.Machine) (bool, error) {
+		return true, nil
+	})
+	// "run the peergrouper" and give all the controllers the vote
+	for _, id := range []string{"0", "1", "2"} {
+		m, err := s.State.Machine(id)
+		c.Assert(err, jc.ErrorIsNil)
+		m.SetHasVote(true)
+	}
+	s.progressControllerToDead(c, "0")
+	// If m0 was 'available' we would be trying to promote it (which would be wrong), but as long as it looks like it
+	// isn't, we should try to replace it
+	// If the Peergrouper had run, then we should have gone down to a single voting machine, but for now we'll do this
+	// in absence of the peergrouper
+	s.assertControllerInfo(c, []string{"1", "2"}, []string{"1", "2"}, nil)
+	changes, err = s.State.EnableHA(3, constraints.Value{}, "quantal", nil, "1")
```

Comment:
This is why...

In order to test that my *test* was correct, I started with a forced count of 3 here, and 5 later on.
These need to be moved back to being 0 so that we trigger the autodetect code, and then we can observe that the computation of how many is correct.

I'm *very* happy that you caught this. Thanks!